### ISSUE TITLE
form metadata cannot be parsed on 0.6 systems

### ISSUE THREAD

**Issue by btbonval** [medic/medic-webapp#1798]
Due to the switch from forms defined in JSON to forms defined in XML to support Enketo, the existing code to interpret couch data into one view per form cannot work for 0.6.

Here is the code which currently parses 0.4 JSON form metadata into relevant views:
https://github.com/medic/middleware/blob/master/postgres/formviews.sql#L37-L43

It expects a ddoc to contain `app_settings.forms.<formname>.fields`, which are currently a dictionary/object/hash by field name. The field name (key) is still required, obviously.

It expects `app_settings.form.<formname>.fields.<fieldname>.labels.short.en` to contain the field's English name as text. This is used as the column label for the form field in the form's view.

It expects `app_settings.form.<formname>.fields.<fieldname>.position` to have an integer. This integer sorts the fields so that the form view builder deterministically generates the same columns in the same order every time for any given form view. This is surprisingly important.

**Comment by btbonval** [comment]
Ideally we can reflect the XML structure directly into the app settings for backwards compatibility. Alternatively we create some novel JSON output and maintain two form metadata parsers in postgres: one for 0.4 and one for 0.6.

**Comment by btbonval** [comment]
I will need help from @alxndrsn or someone familiar with the existing XML parsing in the webapp. I am told there is already a place wherein the XML is parsed in the webapp to support the in app views and this solution to this issue would operate in a similar manner.

**Comment by alxndrsn** [comment]
Do you need this info to come from the design doc, or are you able to query couch by HTTP if the info were available?

**Comment by btbonval** [comment]
We had discussed this with @browndav previously and I agree with him on this point: we want all data processed by Postgres to come from couch documents.

The whole couch database is downloaded as a unit, but making independent HTTP calls could begin to introduce a loss of synchronization between the dataset currently inside postgres and anything fetched by way of HTTP.

**Comment by btbonval** [comment]
I think I worded that poorly. I suppose the couch database is also an HTTP fetch.

The couch database HTTP call is a single atomic fetch that gives a snapshot. Ideally postgres would operate on a single snapshot of the data to maintain integrity, without resorting to data outside that snapshot which might not be in sync with the snapshot.

**Comment by alxndrsn** [comment]
OK.  Have you got an example dataset that you'd expect to see?  I see it's described in the OP, but a concrete example would really help.

**Comment by btbonval** [comment]
I'll see if I can find something in the 0.4 demo dataset and get back to you with highlights from that.

**Comment by btbonval** [comment]
I'm not sure why `forms.js` is separate from `app-settings.js` when in situ the forms are part of app settings. So maybe something has changed with how these things are organized? I don't know if the meta section is required for pulling the `forms.py` contents into `app_settings.forms` indexed by form name or what.

At any rate, here is an example of a P form defined for ANC (part of a list of other forms):
https://github.com/medic/medic-data/blob/1c95974915ee3920c84509020f2d8a542cd159f7/data/generic-anc/base/forms.json#L1-L56

Now here is what that same object might look like with the minimum required for backwards compatibility with the 0.4 postgres form view generator:
```
[{"P":
    {
	"fields": {
		"last_menstrual_period": {
			"labels": {
				"short": {
					"en": "Weeks since LMP"
				}
			},
			"position": 0
		},
		"patient_name": {
			"labels": {
				"short": {
					"en": "Name"
				}
			},
			"position": 0,
		}
	},
  }
},
...
]
```

**Comment by btbonval** [comment]
btw notice how both say `position: 0`? That's technically a bug. I filed it some time ago.
https://github.com/medic/medic-webapp/issues/1308

I could have picked a better example, but `patient_name.position` might instead be 1 to make the above example useful.

**Comment by alxndrsn** [comment]
We already have code to extract the output fields from an XML form submission in `EnketoTranslationService.nodesToJs()` at https://github.com/medic/medic-webapp/blob/develop/static/js/services/enketo-translation.js#L260.  It looks like this code may return a hierarchy in some cases, which I guess would need flattening in a standardised way.  The  `<model>` node from the form definition would normally have the same structure as that in the form submission.

Text labels are derived as `report.{{form_name}}.{{field_name}}`, e.g. `report.pregnancy.patient_name`.

I am not 100% sure where the best place for this data processing is to be done, either for your convenience or where it fits best into `webapp`/`api` (if that is where it must be done).

**Comment by btbonval** [comment]
Dave thinks this should take place in analytics software, so like the data import script.

`ddoc.version` can be used to decide 0.4 and 0.6.

That should be pretty straight  forward at this point for me.

**Comment by btbonval** [comment]
0.4 systems are fine. Easiest approach is to output in a format which matches them so postgres scripts don't need to fool around. Then the conversion process can be built into the importer.

**Comment by btbonval** [comment]
I hit a blocker downloading the dataabase, so I tried to find the XML by looking at the design docs for LG in Futon.

I can only find JSON form definitions.

I saw in medic-projects that each form is defined in XML independently. Somehow my recollection is of being shown a single monolithic XML config for all forms.

Regardless, I cannot find any XML in a live database.

**Comment by alxndrsn** [comment]
I will email some examples.

**Comment by btbonval** [comment]
Thanks @alxndrsn . So that appears to come from the forms view.

I'll see if I can figure out how the forms view generates that internally to figure out where the XML lives in the design doc.

**Comment by btbonval** [comment]
Only thing I can find referring to `/forms` is defined here, but it is for config.
https://github.com/medic/medic-webapp/blob/265b888e3ca4d74c0c04a2649364890238f125e1/static/js/app.js#L238-L246

I can't find `_view/forms` in rewrites or anywhere else.

I tried a different tack searching for XML, but all that comes up is Enketo specific functions which, to the best I can tell, operate on client-side forms.

I can't find where `_view/forms` produces the XML nor where the XML is stored in the ddoc. After 30 mins searching, I feel like I'm better off waiting for someone to tell me.

**Comment by btbonval** [comment]
Honestly, I'm seeing forms defined in app settings and there are field definitions and everything I would expect. I'm not sure why I cannot use those, or what make the XML more authoritative. Marc had explained it but I'm lost in code and data and lack the context to store the things I'm told meaningfully.

At this point I'd be happy just to get the LG ddoc in my database (currently requires a feature to deal with too much data download in the adapter) and see what happens from there.

**Comment by btbonval** [comment]
oh right. I remember now. Those JSON form defs in the LG app_settings are a vestigial organ, like the human appendix. Utility unknown, but left sitting around to confuse people.

A fresh install using 0.6 would presumably have no such thing anymore.

so my best bet is still to get the full raw doc contents into postgres and search for some known XML from the app config within postgres to determine precisely where the form's XML is hiding in the internal couch database.

**Comment by alxndrsn** [comment]
> determine precisely where the form's XML is hiding in the internal couch database after it has been translated into postgres

I don't know where it will be in postgres, but XML forms are defined entirely within their own separate docs.  As you note, the JSON form definition in the ddoc is unrelated.

You can tell an XML form because each doc will have `"type": "form"` and an attachment labelled `xml`.

**Comment by btbonval** [comment]
Oh they are not part of the design doc. That is information I can use and explains ...


everything.

Okay once I get the full dataset loaded in, I'll see if I can find the form defs.

**Comment by btbonval** [comment]
From the import of LG, 11 docs have type form. All 11 have the same top level properties, including `_attachments`. Under that is an `xml` property.
```
bryan=> SELECT count(*) FROM lg WHERE data->>'type' = 'form';
 count 
-------
    11
(1 row)
bryan=> SELECT jsonb_object_keys(data) AS field, count(data) AS docs FROM lg WHERE data->>'type' = 'form' GROUP BY field;
    field     | docs 
--------------+------
 _attachments |   11
 _rev         |   11
 title        |   11
 _id          |   11
 type         |   11
 internalId   |   11
 context      |   11
(7 rows)
bryan=> SELECT count(data->>'_attachments') FROM lg WHERE data->>'type' = 'form' AND data->'_attachments' ? 'xml';
 count 
-------
    11
(1 row)
```

Problem is the xml property only contains metadata:
```
bryan=> SELECT (data->>'_attachments') FROM lg WHERE data->>'type' = 'form' AND data->'_attachments' ? 'xml' LIMIT 1;
 {"xml": {"stub": true, "digest": "md5-qpB0JYZ2SghbmpzWrJVi7g==", "length": 66246, "revpos": 184, "content_type": "text/xml"}}
```

I have no idea where the actual XML is, but I do know it's md5 digest and string length!

**Comment by btbonval** [comment]
Interesting. Trying to see what I got in this database in terms of type. 11000 have no type. Those are placeholders for audit_records which were skipped later on in the import process.

Anyway, I'm not seeing any obvious place to look for XML outside of the form docs.
```
bryan=> SELECT data->>'type' AS type, count(data) AS docs FROM lg GROUP BY type;
       type        | docs  
-------------------+-------
 audit_record      |  5692
 clinic            |  1707
 data_record       |  2569
 district_hospital |     8
 feedback          |   360
 form              |    11
 health_center     |    55
 meta              |     1
 person            |  3596
 usage_stats       |     7
 user-settings     |    50
                   | 11303
(12 rows)
```

**Comment by btbonval** [comment]
Full text search for "xml", "xml version", and "patient_age_in_years". The XML defining forms in medic-projects all starts with `<?xml version=`, but 0 instances of "xml version" exist in the database as I grabbed in. "patient_age_in_years" is probably returning the forms with that field rather than the definition. It is only found in `data_record`s and `audit_record`s.
```
bryan=> SELECT count(data) FROM lg WHERE data::text LIKE '%xml%';
 count 
-------
  3652
(1 row)
bryan=> SELECT count(data) FROM lg WHERE data::text LIKE '%xml version%';
 count 
-------
     0
(1 row)
bryan=> SELECT count(data) FROM lg WHERE data::text LIKE '%patient_age_in_years%';
 count 
-------
  2189
(1 row)
bryan=> SELECT distinct(data->>'type') FROM lg WHERE data::text LIKE '%patient_age_in_years%';
   ?column?   
--------------
 audit_record
 data_record
(2 rows)
```

**Comment by btbonval** [comment]
Also no hit on "bind nodeset", which is part of the XML form definition in medic projects as `<bind nodeset="...`.
```
bryan=> SELECT count(data->>'type') FROM lg WHERE data::text LIKE '%bind nodeset%';
 count 
-------
     0
(1 row)
```

**Comment by btbonval** [comment]
> If we use the ?attachments=true HTTP option when requesting this document, we’d get a Base64-encoded string containing the attachment data. 
http://guide.couchdb.org/draft/api.html#attachments
Maybe that works for bulk requests, in which case this should be fairly easy to sort.

**Comment by btbonval** [comment]
okay, imported again including `?attachments=true`.
```
bryan=> SELECT (data->>'_attachments') FROM lg WHERE data->>'type' = 'form' AND data->'_attachments' ? 'xml' LIMIT 1;
 {"xml": {"data": "PD94bWwgdmVyc2l ...
[tons of data]
```

Unfortunately it isn't decoding from base64 very nicely. I got a bunch of hex values rather than XML ascii when using `SELECT decode( ((data->'_attachments')->'xml')->>'data', 'base64' ) FROM lg WHERE data->>'type' = 'form' AND data->'_attachments' ? 'xml' LIMIT 1;`.

**Comment by btbonval** [comment]
meh. got it to decode part in python, so the base64 is legit and I see xml.
```
>>> base64.b64decode('PD94bWwgdmVyc2lvbj0iMS4wIj8+CjxoOmh0bWwgeG1sbnM9Imh0');
'<?xml version="1.0"?>\n<h:html xmlns="ht'
```

I SEE XML!

**Comment by btbonval** [comment]
Okay so plan of action for the adapter thing:

1. after import is done, determine if version is 0.4 (no action) or 0.6 (continue below)
1. seek out docs with type:form that have `_attachment:xml` property which aren't already in some special place in postgres
1. gank `_attachment:xml:data` for each unparsed doc
1. parse XML for ... I don't actually know yet, write that into postgres somewhere
1. make sure parsed docs are noted so they can be skipped in the above step

**Comment by btbonval** [comment]
Version is in `ddoc.app_settings._version` for 0.4

**Comment by btbonval** [comment]
Version is in `ddoc.kanso.config.version` for 0.6

**Comment by btbonval** [comment]
actually my 0.4 VM also has `ddoc.kanso.config.version` but it shows 0.5

**Comment by btbonval** [comment]
I found other instances of "0.4" in the VM ddoc and "0.6" in the LG ddoc, but they were part of export definitions written into code, not as part of directly readable JSON.

**Comment by btbonval** [comment]
I cannot write integration tests for distinguishing 0.4 and 0.6, or any integration code relying on 0.6, due to lack of a medic-data image or medic-os VM for 0.6.

I can write unit tests though. Do what I can to get this moving.

**Comment by btbonval** [comment]
For parsing XML, the first tag (and it's id) after `<instance>` names the form. Nested inside that will be an `<inputs>` tag with generic fields: `<patient>` tag with tags and content under that. The location meta seems to largely be empty. `<g_inputs>` seems to be empty though the fields look suspiciously like patient fields. `<outputs>` begins a bunch of form-specific fields. Then there are slapdash tags, still under `<formname>` which appear to be calculated? Then there are `<group_*>` tags which appear to contain even more form-specific fields.

The structure seems super arbitrary and unpredictable, but it appears the primary fields of concern for any form are completely written (and nested or not) throughout `<formname>`. The correct approach here might be to flatten all the nesting with underscores with each leaf being a form field (and thus a column at some point). For example, `inputs_patient__id` would be `<inputs><patient><_id>`

```xml
<instance>
  <formname id="formname" version="date">
    <inputs>
      <meta>
        <location>
          <lat/>
          <long/>
         </location>
       </meta>
       <patient>
         <_id>UUID</_id>
         <name>person's name</name>
         <date_of_birth>date</date_of_birth>
       </patient>
     </inputs>
     <g_inputs>
       <_id/>
       <name/>
       <date_of_birth/>
    </g_inputs>
    <outputs>
      <stuff>things</stuff>
    </outputs>
    <random_thing/>
    <different_random_thing>with contents!</different_random_thing>
    <group_bananas>
      <useful_tag/>
      <question>answer</question>
    </group_bananas>
  </formname>
</instance>
```

**Comment by btbonval** [comment]
Empty leaf tags should probably take `NULL`.

**Comment by btbonval** [comment]
Okay so just for good measure, let's convert the above to JSON by hand. Heck might make a good tests fixture. I'm missing nice names, but for now I'll just run with the flattened name in all cases. Position can be an integer which simply increments across tags in their proper order.

I'm going to flatten with spaces b/c why not? They're just strings in a database. Make it look like 0.4 form defs:
```json
[{"formname":
    {
    "fields": {

        "inputs meta location lat": {
            "labels": {
                "short": {name
                    "en": "inputs meta location lat"
                }
            },
            "position": 0
        },
        "inputs meta location long": {
            "labels": {
                "short": {
                    "en": "inputs meta location long"
                }
            },
            "position": 1
        },
        "inputs patient _id": {
            "labels": {
                "short": {
                    "en": "inputs patient _id"
                }
            },
            "position": 2
        },
        "inputs patient name": {
            "labels": {
                "short": {
                    "en": "inputs patient name"
                }
            },
            "position": 3
        },
        "inputs patient date_of_birth": {
            "labels": {
                "short": {
                    "en": "inputs patient date_of_birth"
                }
            },
            "position": 4
        },
        "g_inputs _id": {
            "labels": {
                "short": {
                    "en": "g_inputs _id"
                }
            },
            "position": 5
        },
        "g_inputs name": {
            "labels": {
                "short": {
                    "en": "g_inputs name"
                }
            },
            "position": 6
        },
        "g_inputs date_of_birth": {
            "labels": {
                "short": {
                    "en": "g_inputs date_of_birth"
                }
            },
            "position": 7
        },
        "outputs stuff": {
            "labels": {
                "short": {
                    "en": "outputs stuff"
                }
            },
            "position": 8
        },
        "random_thing": {
            "labels": {
                "short": {
                    "en": "random_thing"
                }
            },
            "position": 9
        },
        "different_random_thing": {
            "labels": {
                "short": {
                    "en": "different_random_thing"
                }
            },
            "position": 10
        },
        "group_bananas useful_tag": {
            "labels": {
                "short": {
                    "en": "group_bananas useful_tag"
                }
            },
            "position": 11
        },
        "group_bananas question": {
            "labels": {
                "short": {
                    "en": "group_bananas question"
                }
            },
            "position": 12
        },
    },
  }
},
...
]
```

**Comment by btbonval** [comment]
I just realized that the actual reports from these forms (type:data_record) are maintaining the form submissions in XML as well. There are a few token fields in JSON, but it looks like if I am to access anything I need, I must parse not only the form definitions, but also the forms themselves, using XML. This is a major complication.

**Comment by alxndrsn** [comment]
> I'm going to flatten with spaces b/c why not? They're just strings in a database.

If you separate with `/`, you might find it handy later if you need to query form submission models using XPath.

**Comment by btbonval** [comment]
@alxndrsn The field names will be columns of tables when flattened, so I should be querying tables directly using SQL right?

I might need to look into XPath or some kind of node XML parser to parse the trees in the first place, however.

**Comment by alxndrsn** [comment]
> The field names will be columns of tables when flattened, so I should be querying tables directly using SQL right?
Ah yes, so as long as the node to column name conversion is consistent between both forms and reports, I guess it won't make a difference.

**Comment by btbonval** [comment]
Form definitions hide the XML as noted above in an attachment. https://github.com/medic/medic-webapp/issues/1798#issuecomment-179074125

Actual form reports are not attachments. Look for `type:data_record` and `form:something`, and then it appears the XML is not even base64 encoded as form defs are. It's just `doc.content` as a string.

**Comment by btbonval** [comment]
Postgres has XML support which is basically limited to XPath for parsing and cannot be optimized by indices.
* http://www.postgresql.org/docs/9.4/static/datatype-xml.html
* http://www.postgresql.org/docs/9.4/static/functions-xml.html

There exist XPath definitions to find all leaf tags and their contents, however, they do not return the full path so like-named tags in different hierarchical places could collide.
* http://stackoverflow.com/questions/3926589/how-to-select-all-leaf-nodes-using-xpath-expression

There exist XSLTs which transform XML into its leaf node XPaths. Postgres does not support XSLT, but Node does.
* http://stackoverflow.com/questions/9064951/how-to-get-xpaths-for-all-leaf-elements-from-xml
* https://www.npmjs.com/package/node_xslt

The form definition could be read into Node (trimming to just the `<instance>`), then transformed by the 3rd party XSLT to return only the leaf elements as a list of XPaths per form which get stored ?somewhere?. Postgres could have an additional table for storing miscellania outside of Couch.

Node would then need to read all form reports, look up the XPaths for the given form, and return the set of XPath results as a row. Field names in a view might as well be XPaths as Alex suggested.

**Comment by btbonval** [comment]
Report parsing will probably require some kind of pagination to manage indefinite numbers of collected reports which need parsing. Some kind of comparison for already-parsed reports will be helpful.

**Comment by mandric** [comment]
@btbonval use the app-settings API for now...  until we have a medic-api endpoint.
```
# get medic-webapp version 
curl https://username:secret@hostname/medic/_design/medic/_rewrite/app_settings/medic | \
  jq .meta.config.version
```

**Comment by garethbowen** [comment]
... or ddoc in `kanso.config.version`

**Comment by btbonval** [comment]
@garethbowen `kanso.config.version` shows 0.5 in 0.4 medic-os release which I've been using for local VM testing. It is from July, but I don't think a newer 0.4 medic-os has been released.

**Comment by garethbowen** [comment]
You mean medic-webapp version, not medic-os version, right?

There was a 0.5 version which IIRC was never fully released, which is why we usually only talk about 0.4 and 0.6.

**Comment by btbonval** [comment]
@garethbowen no, I mean medic-os. I use a VM based on medic-os-v1.1.0-20150731-x86-virtual.iso downloaded from dev.medicmobile.org

**Comment by btbonval** [comment]
Sounds like `kanso.config.version` is most canonical and indeed the VM I'm running is a 0.5 version.

**Comment by btbonval** [comment]
okay so it seems like I don't need to check for version. XML forms will be `type:form` containing non-empty `_attachments.xml`, while non-XML forms will be ... something else. 0.4 won't even have any such docs, so merely searching for them and iterating will skip the XML part as a matter of course.

**Comment by btbonval** [comment]
Unit tests are to the point where (for form definitions), XML is being converted into a flattened list of field names, so like: `{ "formname": [ "field1", "field2", "category/field3", "category/field4" ] } `.

To maintain some backwards compatibility with existing calculations, the formnames must be entered into a table (view won't cut the mustard due to XML) holding the name of all forms. Need to create the table if it doesn't exist as `formlist (form TEXT)` and do smart insertions using the pg-promise current best practices.

Then a table must be created for each form as `formview_{name} (field1 TEXT, field2 TEXT, category/field3 TEXT, category/field4 TEXT)`. The current 0.4 stuff pretty much assumes all fields are text and casts when necessary, so text fields will work fine. Views are fortunate enough not to be typed.

Later in the process, smart insertions into the above `formview_{name}` tables will be required, but it should be pretty darn similar code.

All insertions should use current pg-promise best practices.  https://github.com/vitaly-t/pg-promise/wiki/Performance-Boost

**Comment by btbonval** [comment]
rolling over to current iteration because it's not done. started it mid-iteration if not later.

**Comment by btbonval** [comment]
Limitation: due to creating tables out of these forms in order to keep track of what has already been downloaded (iterative update) and manage some other problems (can't parse XML in pg), a form cannot have dynamic fields.

Even if I were to do something a little more flexible where I dump forms as JSON and create a VIEW as the current 0.4 postgres stuff works, the VIEW will have restricted fields and won't necessarily make sense of new fields on old records of the same type.

I think it's safe to say changing fields on an existing form is a bad idea (tm) regardless.

**Comment by btbonval** [comment]
I'm trying to run the form extraction code against a postgres database. In a database with no XML forms, the code gets all the way to the end and then hangs. That was very much like this problem: https://github.com/medic/medic-webapp/issues/1799#issuecomment-176482512

I can't seem to resolve the hang, it's a different issue I need to poke at more.

When trying the code against a database with forms, it picks out the expected 11 forms and tries to extract instances, but then skips to the last step and hangs. I'm guessing an error occurred, but I don't understand why `.then(function (), function (err) {report err})` is failing to trap errors. I need to poke at this more as well.

**Comment by btbonval** [comment]
I don't know why error handling doesn't work as advertised with RSVP promises. I had to put an explicit `catch()` after the exact `then()` which is erroring. A `then(good(), bad())` was not catching anything, and a `.catch()` at the very end didn't pick anything up. It had to be `.then(good).catch(bad)`.

Error for 0.6:
```
[Error: Parse Error: Malformed tag on line 184: <bind calculate="if ( /assessment/group_cough/patient_coughs ='yes', 
 concat (
 'Cough', 
 if( string-length( /assessment/group_cough/coughing_duration ) != 0,
 concat (
 ' for ', 
 jr:choice-name( /assessment/group_cough/coughing_duration , ' /assessment/group_cough/coughing_duration ')
 ),
 ''
 )
 ),
 ''
)" nodeset="/assessment/outputs/symptom_cough" type="string"/>]
```

**Comment by btbonval** [comment]
The `then`/`catch` code was fine. Actually the problem was my error handler: it was `throw`ing the error (which was silently ignored) but did not send anything to log. Logging added, this helps.

**Comment by btbonval** [comment]
The 0.4 failure to close was just impatience. It will close out of its own accord, it just takes a great many seconds.

So it looks like the only problem stopping my form definition import is form def XML that cannot be parsed.

**Comment by btbonval** [comment]
So forms are just going to change willy nilly and somehow I have to make the database work with that.

I'll need to include tracking of the version property of the tag. It sounds like the old definitions will simply be gone?

We can add some kind of configuration or something to manage joining a similar type of form across versions of that form in analytics later.

**Comment by btbonval** [comment]
create a table `formview_formname_version` for each version as they are encountered, populate unread data into the "current" version, create a view for `formview_formname_X` possibly built dynamically by a function.

**Comment by btbonval** [comment]
Need to capture reported date of forms, not part of the XML, but is part of the JSON metadata, and associate it with the correct form.

**Comment by btbonval** [comment]
Need to add a table for tracking additional CHP information. UUID string primary key, additional fields include active/inactive boolean. Ideally UUID would be a FK against a CHP table, but there won't be a CHP table, there will be a contacts table with a CHP view. So things get a little wonky.

**Comment by btbonval** [comment]
Contacts are probably found in four docs with these UUIDs:
```
 form:contact:clinic:create
 form:contact:clinic:edit
 form:contact:person:create
 form:contact:person:edit
```

**Comment by alxndrsn** [comment]
@btbonval by convention any form whose ID starts `form:contact:` is for creating/editing contacts rather than for a report.

**Comment by btbonval** [comment]
Problem: I can get the most recent form definitions (all currently versioned on 2015-11-23), but I cannot get the old versions.

I don't currently have an easy way to test if a table already exists for a given form and version at any time, especially not across iterative runs. I need to track that somehow. If a table exists, I can proceed with data processing. If not, I should not add that data but I should mark it somehow.

On that note, I need to come up with a view which shows me useful form_metadata for reports which are in form_metadata but not in any other table.

**Comment by btbonval** [comment]
** There's some internal pg table thing that contains all tables. I can query against that to see which tables exist given a list of tables I might have data for. Need to find that thing.

**Comment by browndav** [comment]
Take a look at http://www.postgresql.org/docs/current/static/information-schema.html

**Comment by btbonval** [comment]
Thanks @browndav . Off my memory I think it's just in `pg_catalog` somewheres, there's a stackoverflow for it. It was getting on 4:30am so I wanted to make a note to look it up before I forgot.

After sleeping on it, I think I'm going to give up on parsing form definitions, and rather, read the form reports themselves and use them to generate tables as needed. I'm only getting the most recent version of any form definition, but the form reports sometimes span to previous versions which I haven't created a table for. In that way, I don't need to check whether or not a table exists, I can just `CREATE TABLE ... IF NOT EXISTS` for each table required as discovered from form reports.

**Comment by btbonval** [comment]
Basically done now. Tested using my trial LG database and was able to generate more than 20 tables for each form/version combination present in the data, and populate those tables with the records!

Things:

* I saw `uuid:LONG_HEXADECIMAL_THING` in a few places, not sure why `uuid:` got prepended though. In other places, it's just the hexadecimal.
* I need to test partial updates.
* I need to clean up old method of using form definitions, as now I just use the XML of the form reports themselves to build the definitions, bypassing a ton of code.

Once the above is done, onto contacts.

**Comment by btbonval** [comment]
Hand testing partial updates:

* Base step sucked in all available 2571 docs.
* Run after that sucks in 0 docs. (as expected).
* Deleted 786 reports from metadata and pregnancy 2015-11-23 forms. Ran. Sucks in 786 docs. metadata and pregnancy 2015-11-23 checked, got all 786 back again.

Wrote and pushed SQL tests. Cleaned up old form definition approach and pushed.

I'm not immediately concerned about the `uuid:` prefix. I mean I am, but not for this ticket. It might become a thing later for dealing with contacts. That sounds like a contact issue problem though.

Closing as done.
